File: cli/src/tools/filesystem.ts
/**
 * File system tools
 */

import { promises as fs } from 'fs';
import { join, dirname } from 'path';
import { ToolResult, ok, err } from './index.js';

/**
 * Read file contents
 */
export async function readFile(path: string): Promise<ToolResult> {
  try {
    const content = await fs.readFile(path, 'utf-8');
    return ok(content);
  } catch (error: any) {
    return err(`Failed to read file: ${error.message}`);
  }
}

/**
 * Write file contents
 */
export async function writeFile(path: string, content: string): Promise<ToolResult> {
  try {
    // Create parent directories if needed
    const dir = dirname(path);
    await fs.mkdir(dir, { recursive: true });

    await fs.writeFile(path, content, 'utf-8');
    return ok(`Written ${content.length} bytes to ${path}`);
  } catch (error: any) {
    return err(`Failed to write file: ${error.message}`);
  }
}

/**
 * List directory contents
 */
export async function listDir(path: string, recursive = false): Promise<ToolResult> {
  try {
    const stat = await fs.stat(path);
    if (!stat.isDirectory()) {
      return err(`Not a directory: ${path}`);
    }

    const entries: string[] = [];

    async function walkDir(dir: string, prefix: string): Promise<void> {
      const items = await fs.readdir(dir, { withFileTypes: true });
      for (const item of items) {
        const displayPath = prefix ? `${prefix}/${item.name}` : item.name;
        if (item.isDirectory()) {
          entries.push(`${displayPath}/`);
          if (recursive) {
            await walkDir(join(dir, item.name), displayPath);
          }
        } else {
          entries.push(displayPath);
        }
      }
    }

    await walkDir(path, '');
    entries.sort();
    return ok(entries.join('\n'));
  } catch (error: any) {
    return err(`Failed to list directory: ${error.message}`);
  }
}

/**
 * Copy file
 */
export async function copyFile(src: string, dst: string): Promise<ToolResult> {
  try {
    const srcStat = await fs.stat(src);
    if (!srcStat.isFile()) {
      return err('Source is not a file');
    }

    // Create parent directories
    const dir = dirname(dst);
    await fs.mkdir(dir, { recursive: true });

    await fs.copyFile(src, dst);
    return ok(`Copied ${src} to ${dst}`);
  } catch (error: any) {
    return err(`Failed to copy: ${error.message}`);
  }
}

/**
 * Move/rename file or directory
 */
export async function movePath(src: string, dst: string): Promise<ToolResult> {
  try {
    await fs.rename(src, dst);
    return ok(`Moved ${src} to ${dst}`);
  } catch (error: any) {
    return err(`Failed to move: ${error.message}`);
  }
}

/**
 * Delete file or directory
 */
export async function deletePath(path: string, force = false): Promise<ToolResult> {
  try {
    const stat = await fs.stat(path);

    if (stat.isDirectory()) {
      if (force) {
        await fs.rm(path, { recursive: true, force: true });
        return ok(`Deleted directory: ${path}`);
      } else {
        return err('Use force=true to delete directories');
      }
    } else {
      await fs.unlink(path);
      return ok(`Deleted file: ${path}`);
    }
  } catch (error: any) {
    return err(`Failed to delete: ${error.message}`);
  }
}

/**
 * Get file metadata
 */
export async function fileInfo(path: string): Promise<ToolResult> {
  try {
    const stat = await fs.stat(path);

    const fileType = stat.isDirectory() ? 'directory' : stat.isFile() ? 'file' : 'other';
    const modified = stat.mtime.toISOString();

    return ok(
      `Type: ${fileType}\n` +
      `Size: ${stat.size} bytes\n` +
      `Modified: ${modified}\n` +
      `Read-only: ${!(stat.mode & 0o200)}`
    );
  } catch (error: any) {
    return err(`Failed to get file info: ${error.message}`);
  }
}

/**
 * Search for files by pattern
 */
export async function findFiles(
  directory: string,
  pattern: string,
  maxResults = 100
): Promise<ToolResult> {
  try {
    const { globby } = await import('globby');
    const results = await globby(pattern, {
      cwd: directory,
      gitignore: true,
    });

    const limited = results.slice(0, maxResults);
    if (results.length > maxResults) {
      return ok(limited.join('\n') + `\n... and ${results.length - maxResults} more`);
    }
    return ok(limited.join('\n') || '(no matches)');
  } catch (error: any) {
    return err(`Failed to search: ${error.message}`);
  }
}

/**
 * File system tool definitions for agents
 */
export const filesystemTools = [
  {
    name: 'read_file',
    description: 'Read file contents',
    parameters: {
      type: 'object',
      properties: {
        path: { type: 'string', description: 'File path' },
      },
      required: ['path'],
    },
  },
  {
    name: 'write_file',
    description: 'Write content to file',
    parameters: {
      type: 'object',
      properties: {
        path: { type: 'string', description: 'File path' },
        content: { type: 'string', description: 'Content to write' },
      },
      required: ['path', 'content'],
    },
  },
  {
    name: 'list_dir',
    description: 'List directory contents',
    parameters: {
      type: 'object',
      properties: {
        path: { type: 'string', description: 'Directory path' },
        recursive: { type: 'boolean', description: 'Include subdirectories' },
      },
      required: ['path'],
    },
  },
  {
    name: 'copy_file',
    description: 'Copy file',
    parameters: {
      type: 'object',
      properties: {
        src: { type: 'string', description: 'Source path' },
        dst: { type: 'string', description: 'Destination path' },
      },
      required: ['src', 'dst'],
    },
  },
  {
    name: 'move_path',
    description: 'Move or rename file/directory',
    parameters: {
      type: 'object',
      properties: {
        src: { type: 'string', description: 'Source path' },
        dst: { type: 'string', description: 'Destination path' },
      },
      required: ['src', 'dst'],
    },
  },
  {
    name: 'delete_path',
    description: 'Delete file or directory',
    parameters: {
      type: 'object',
      properties: {
        path: { type: 'string', description: 'Path to delete' },
        force: { type: 'boolean', description: 'Force delete directories' },
      },
      required: ['path'],
    },
  },
  {
    name: 'file_info',
    description: 'Get file metadata',
    parameters: {
      type: 'object',
      properties: {
        path: { type: 'string', description: 'File path' },
      },
      required: ['path'],
    },
  },
  {
    name: 'find_files',
    description: 'Search for files by glob pattern',
    parameters: {
      type: 'object',
      properties: {
        directory: { type: 'string', description: 'Search directory' },
        pattern: { type: 'string', description: 'Glob pattern (e.g., "**/*.ts")' },
      },
      required: ['directory', 'pattern'],
    },
  },
];

File: src/tools/filesystem.rs
//! File system tools

use super::ToolResult;
use std::path::Path;

/// Read file contents
pub fn read_file(path: &str) -> ToolResult {
    match std::fs::read_to_string(path) {
        Ok(content) => ToolResult::ok(content),
        Err(e) => ToolResult::err(format!("Failed to read file: {}", e)),
    }
}

/// Write file contents
pub fn write_file(path: &str, content: &str) -> ToolResult {
    // Create parent directories if needed
    if let Some(parent) = Path::new(path).parent() {
        if let Err(e) = std::fs::create_dir_all(parent) {
            return ToolResult::err(format!("Failed to create directory: {}", e));
        }
    }

    match std::fs::write(path, content) {
        Ok(()) => ToolResult::ok(format!("Written {} bytes to {}", content.len(), path)),
        Err(e) => ToolResult::err(format!("Failed to write file: {}", e)),
    }
}

/// List directory contents
pub fn list_dir(path: &str, recursive: bool) -> ToolResult {
    let path = Path::new(path);

    if !path.exists() {
        return ToolResult::err(format!("Path does not exist: {}", path.display()));
    }

    if !path.is_dir() {
        return ToolResult::err(format!("Not a directory: {}", path.display()));
    }

    let mut entries = Vec::new();

    if recursive {
        fn walk_dir(dir: &Path, entries: &mut Vec<String>, prefix: &str) -> std::io::Result<()> {
            for entry in std::fs::read_dir(dir)? {
                let entry = entry?;
                let path = entry.path();
                let name = entry.file_name().to_string_lossy().to_string();
                let display = if prefix.is_empty() {
                    name.clone()
                } else {
                    format!("{}/{}", prefix, name)
                };

                if path.is_dir() {
                    entries.push(format!("{}/", display));
                    walk_dir(&path, entries, &display)?;
                } else {
                    entries.push(display);
                }
            }
            Ok(())
        }

        if let Err(e) = walk_dir(path, &mut entries, "") {
            return ToolResult::err(format!("Failed to read directory: {}", e));
        }
    } else {
        match std::fs::read_dir(path) {
            Ok(dir) => {
                for entry in dir.flatten() {
                    let name = entry.file_name().to_string_lossy().to_string();
                    if entry.path().is_dir() {
                        entries.push(format!("{}/", name));
                    } else {
                        entries.push(name);
                    }
                }
            }
            Err(e) => return ToolResult::err(format!("Failed to read directory: {}", e)),
        }
    }

    entries.sort();
    ToolResult::ok(entries.join("\n"))
}

/// Copy file or directory
pub fn copy(src: &str, dst: &str) -> ToolResult {
    let src_path = Path::new(src);
    let dst_path = Path::new(dst);

    if !src_path.exists() {
        return ToolResult::err(format!("Source does not exist: {}", src));
    }

    if src_path.is_file() {
        // Copy single file
        if let Some(parent) = dst_path.parent() {
            if let Err(e) = std::fs::create_dir_all(parent) {
                return ToolResult::err(format!("Failed to create directory: {}", e));
            }
        }

        match std::fs::copy(src_path, dst_path) {
            Ok(bytes) => ToolResult::ok(format!("Copied {} bytes from {} to {}", bytes, src, dst)),
            Err(e) => ToolResult::err(format!("Failed to copy: {}", e)),
        }
    } else {
        // Copy directory recursively
        ToolResult::err("Directory copy not yet implemented")
    }
}

/// Move/rename file or directory
pub fn move_path(src: &str, dst: &str) -> ToolResult {
    match std::fs::rename(src, dst) {
        Ok(()) => ToolResult::ok(format!("Moved {} to {}", src, dst)),
        Err(e) => ToolResult::err(format!("Failed to move: {}", e)),
    }
}

/// Delete file or directory
pub fn delete(path: &str, force: bool) -> ToolResult {
    let path = Path::new(path);

    if !path.exists() {
        return ToolResult::err(format!("Path does not exist: {}", path.display()));
    }

    if path.is_dir() {
        if force {
            match std::fs::remove_dir_all(path) {
                Ok(()) => ToolResult::ok(format!("Deleted directory: {}", path.display())),
                Err(e) => ToolResult::err(format!("Failed to delete: {}", e)),
            }
        } else {
            ToolResult::err("Use force=true to delete directories")
        }
    } else {
        match std::fs::remove_file(path) {
            Ok(()) => ToolResult::ok(format!("Deleted file: {}", path.display())),
            Err(e) => ToolResult::err(format!("Failed to delete: {}", e)),
        }
    }
}

/// Get file metadata
pub fn file_info(path: &str) -> ToolResult {
    let path = Path::new(path);

    match std::fs::metadata(path) {
        Ok(meta) => {
            let file_type = if meta.is_dir() {
                "directory"
            } else if meta.is_file() {
                "file"
            } else {
                "other"
            };

            let size = meta.len();
            let readonly = meta.permissions().readonly();

            let modified = meta
                .modified()
                .ok()
                .map(|t| {
                    let datetime: chrono::DateTime<chrono::Local> = t.into();
                    datetime.format("%Y-%m-%d %H:%M:%S").to_string()
                })
                .unwrap_or_else(|| "unknown".to_string());

            ToolResult::ok(format!(
                "Type: {}\nSize: {} bytes\nModified: {}\nRead-only: {}",
                file_type, size, modified, readonly
            ))
        }
        Err(e) => ToolResult::err(format!("Failed to get file info: {}", e)),
    }
}
